### PR TITLE
Fix git.Blob.DataAsync(): close pipe since we return a NopCloser (#16899)

### PR DESCRIPTION
**Pull request by 6543**
Backport #16899

* make sure headGitRepo is closed on err too
* refactor
* Fix git.Blob.DataAsync(): exec cancel since we already read all bytes (close pipe since we return a NopCloser)

### CODE REVIEW

File: routers/web/repo/pull.go [hunk 1]
```diff
@@ -1012,10 +1012,14 @@ func CompareAndPullRequestPost(ctx *context.Context) {
 	)
 
 	headUser, headRepo, headGitRepo, prInfo, baseBranch, headBranch := ParseCompareInfo(ctx)
+	defer func() {
```
Comment (zeripath) [inline routers/web/repo/pull.go:1015]: This was fin but I'm not going to argue

File: routers/web/repo/issue.go [hunk 2]
```diff
@@ -419,9 +419,6 @@ func RetrieveRepoMilestonesAndAssignees(ctx *context.Context, repo *models.Repos
 	}
 
 	handleTeamMentions(ctx)
-	if ctx.Written() {
```
Comment (zeripath) [inline routers/web/repo/issue.go:422]: Seems like a bad idea to remove this?

**Review by zeripath — APPROVED** [review]
~~Not sure about the wisdom of the ctx.Written change otherwise~~ it's fine.  lgtm

File: routers/web/repo/issue.go (hunk 2 above)
Comment (zeripath) [reply routers/web/repo/issue.go:422]: it's fine

**Review by lunny — APPROVED** [review]

File: routers/web/repo/pull.go (hunk 1 above)
Comment (6543) [reply routers/web/repo/pull.go:1015]: Yes it is in the refactor commit ...

Make it more redable

**Pull request merged**